internal/protocol/gcs.go

```diff
@@ -48,15 +51,96 @@ func (GcsBucketInfoTranslator) ApiToRpc(vars map[cosiapi.BucketInfoVar]string) *
 		return nil
 	}
 
-	// TODO: implement
+	out := &cosiproto.GcsBucketInfo{}
 
-	return nil
+	out.BucketName = vars[cosiapi.BucketInfoVar_GCS_BucketName]
+	out.ProjectId = vars[cosiapi.BucketInfoVar_GCS_ProjectId]
+
+	return out
 }
 
 func (GcsBucketInfoTranslator) Validate(
 	vars map[cosiapi.BucketInfoVar]string, _ cosiapi.BucketAccessAuthenticationType,
 ) error {
-	// TODO: implement
+	errs := []string{}
+
+	bucketName := vars[cosiapi.BucketInfoVar_GCS_BucketName]
+	if bucketName == "" {
+		errs = append(errs, "GCS bucket name cannot be unset")
+	}
+
+	projectId := vars[cosiapi.BucketInfoVar_GCS_ProjectId]
+	if projectId == "" {
+		errs = append(errs, "GCS project ID cannot be unset")
+	}
+
+	if len(errs) > 0 {
+		return fmt.Errorf("GCS bucket info is invalid: %v", errs)
+	}
+	return nil
+}
+
+func (GcsCredentialTranslator) RpcToApi(c *cosiproto.GcsCredentialInfo) map[cosiapi.CredentialVar]string {
+	if c == nil {
+		return nil
+	}
 
+	out := map[cosiapi.CredentialVar]string{
+		cosiapi.CredentialVar_GCS_AccessId:       c.AccessId,
+		cosiapi.CredentialVar_GCS_AccessSecret:   c.AccessSecret,
+		cosiapi.CredentialVar_GCS_PrivateKeyName: c.PrivateKeyName,
+		cosiapi.CredentialVar_GCS_ServiceAccount: c.ServiceAccount,
+	}
+
+	return out
+}
+
+func (GcsCredentialTranslator) ApiToRpc(vars map[cosiapi.CredentialVar]string) *cosiproto.GcsCredentialInfo {
+	if len(vars) == 0 {
+		return nil
+	}
+
+	out := &cosiproto.GcsCredentialInfo{}
+
+	out.AccessId = vars[cosiapi.CredentialVar_GCS_AccessId]
+	out.AccessSecret = vars[cosiapi.CredentialVar_GCS_AccessSecret]
+	out.PrivateKeyName = vars[cosiapi.CredentialVar_GCS_PrivateKeyName]
+	out.ServiceAccount = vars[cosiapi.CredentialVar_GCS_ServiceAccount]
+
+	return out
+}
+
+func (GcsCredentialTranslator) Validate(
+	vars map[cosiapi.CredentialVar]string, authType cosiapi.BucketAccessAuthenticationType,
+) error {
+	errs := []string{}
+
+	switch authType {
+	case cosiapi.BucketAccessAuthenticationTypeKey:
+		accessId := vars[cosiapi.CredentialVar_GCS_AccessId]
+		if accessId == "" {
+			errs = append(errs, "GCS access ID cannot be unset")
+		}
+
+		accessSecret := vars[cosiapi.CredentialVar_GCS_AccessSecret]
+		if accessSecret == "" {
+			errs = append(errs, "GCS access secret cannot be unset")
+		}
+
+	case cosiapi.BucketAccessAuthenticationTypeServiceAccount:
+		privateKeyName := vars[cosiapi.CredentialVar_GCS_PrivateKeyName]
+		if privateKeyName == "" {
+			errs = append(errs, "GCS private key name cannot be unset")
+		}
+
+		serviceAccount := vars[cosiapi.CredentialVar_GCS_ServiceAccount]
+		if serviceAccount == "" {
+			errs = append(errs, "GCS service account cannot be unset")
+		}
+	}
```

Comment:
This is my understanding as per:
https://docs.databricks.com/gcp/en/connect/storage/gcs#step-2-create-a-key-to-access-gcs-bucket-directly
https://docs.snowflake.com/en/connectors/google/gard/gard-connector-create-service-account-key#prerequisites